pkg/controller/util/endpoint/controller_utils.go

```diff
@@ -32,6 +34,99 @@ import (
 // EndpointsMatch is a type of function that returns true if pod endpoints match.
 type EndpointsMatch func(*v1.Pod, *v1.Pod) bool
 
+// ServiceSelectorCache is a cache of service selectors to avoid high CPU consumption caused by frequent calls to AsSelectorPreValidated (see #73527)
+type ServiceSelectorCache struct {
+	cache map[string]labels.Selector
+	lock  sync.RWMutex
+}
+
+// NewServiceSelectorCache init ServiceSelectorCache for both endpoint controller and endpointSlice controller.
+func NewServiceSelectorCache() *ServiceSelectorCache {
+	return &ServiceSelectorCache{
+		cache: map[string]labels.Selector{},
+	}
+}
+
+// Get return selector and existence in ServiceSelectorCache by key.
+func (sc *ServiceSelectorCache) Get(key string) (labels.Selector, bool) {
+	sc.lock.RLock()
+	defer sc.lock.RUnlock()
+	selector, ok := sc.cache[key]
+	return selector, ok
+}
+
+// Update can update selector of ServiceSelectorCache while service's selector changed.
+// Update can also add selector to ServiceSelectorCache while service's selector no cached.
+func (sc *ServiceSelectorCache) Update(key string, service *v1.Service) labels.Selector {
+	sc.lock.Lock()
+	defer sc.lock.Unlock()
+	v, ok := sc.cache[key]
+	// service selector not exist in ServiceSelectorCache, add it to ServiceSelectorCache.
+	if !ok {
+		selector := labels.Set(service.Spec.Selector).AsSelectorPreValidated()
+		sc.cache[key] = selector
+		return selector
+	}
+
+	// service selector exist in ServiceSelectorCache, update it in ServiceSelectorCache if selector changed.
+	oldSelector := v.(labels.Selector)
+	curSelector := labels.Set(service.Spec.Selector).AsSelectorPreValidated()
+	if reflect.DeepEqual(curSelector, oldSelector) {
+		return oldSelector
+	}
+	// service selector changed, so update ServiceSelectorCache
+	sc.cache[key] = curSelector
+	return curSelector
```

Comment:
I think this can all be replaced with 
```
selector := labels.Set(service.Spec.Selector).AsSelectorPreValidated()
sc.cache[key] = selector
return selector
```

Since the expensive thing is generating a selector, and we're doing that anyway as part of a comparison, the comparison doesn't really save us anything. The big gain here is that we're only generating these selectors when a Service changes, not for every related Service every time a Pod changes.